### PR TITLE
refactor mikrotik formatter

### PR DESCRIPTION
**Pull request by j3n57h0m45**
#### Old Formatter
The previous formatter operated sequentially by first dropping the entire blocklist and then repopulating it with new entries. This approach had several drawbacks:

1. **Resource Intensive**: Dropping the blocklist was particularly resource-intensive, especially on lower-end Mikrotik devices.
2. **Security Exposure**: The device was exposed while the blocklist was being repopulated, as there were no entries in the list during this time.
3. **Inefficient Timeout Utilization**: The `timeout` parameter was not utilized effectively in generating the address list.

#### Refactored Formatter
The `mikrotik` formatter has now been refactored into a separate package and uses Go templates to generate the Mikrotik script. The new implementation addresses the previous shortcomings as follows:

1. **No Full Blocklist Drop**: The new script avoids the need to drop the entire blocklist.
2. **Efficient Timeout Usage**: The `timeout` parameter is now properly utilized.
3. **Global Function Usage**: The generated Mikrotik script uses a global function ([available since RouterOS 6.2](https://help.mikrotik.com/docs/display/ROS/Scripting)) to add or modify each entry.

### CODE REVIEW

**Comment by LaurenceJJones** [comment]
So im not a user of microtik just so I understand:

The template creates a reusable function which in short adds and remove within the `do` function and even if it fails it will readd it never the less. The timeout is handled on microtik side and we dont care if a decision was deleted since the timeout was set correctly?

With the go embed couldnt we just embed the file itself instead of using a file system, then we would need to read the file and can expect it to be a string or bytes?

```go
//go:embed mikrotik.tmpl
var MikrotikScriptTemplate string

/* this can be removed
MikrotikScriptTemplate, err := fs.ReadFile(MikrotikScriptTemplateFile, "mikrotik.tmpl")
if err != nil {
	http.Error(w, "Error reading template: "+err.Error(), http.StatusInternalServerError)
	return
}  */

parsedTemplate, err := template.New("script").Parse(MikrotikScriptTemplate)
```

**Comment by j3n57h0m45** [comment]
> The template creates a reusable function which in short adds and remove within the `do` function and even if it fails it will readd it never the less. The timeout is handled on microtik side and we dont care if a decision was deleted since the timeout was set correctly?

yes thats correct. 

> With the go embed couldnt we just embed the file itself instead of using a file system, then we wouldn't need to read the file each time and can expect it to be a string or bytes?

true it makes more sense to just embed it as a string.

**Comment by j3n57h0m45** [comment]
also, if the new implementation is good to be approved, the [docker/README.md](https://github.com/crowdsecurity/cs-blocklist-mirror/blob/83ff2b5b0813d9c8321c21e295525a36182a3442/docker/README.md) needs some polishing.

**Comment by LaurenceJJones** [comment]
i'll see if I can ask @williamdes if the implementation that is used on microtik side seems okay, since we are limited on our side for testing.

**Comment by j3n57h0m45** [comment]
> i'll see if I can ask @williamdes if the implementation that is used on microtik side seems okay, since we are limited on our side for testing.

The generated mikrotik script code has been tested / is in use on a MikroTik hAP ac with RouterOS 7.15.1

**Comment by LaurenceJJones** [comment]
> > i'll see if I can ask @williamdes if the implementation that is used on microtik side seems okay, since we are limited on our side for testing.
> 
> The generated mikrotik script code has been tested / is in use on a MikroTik hAP ac with RouterOS 7.15.1

We should also add to the documentation the min version of RouterOS needed

**Comment by williamdes** [comment]
> i'll see if I can ask @williamdes if the implementation that is used on microtik side seems okay, since we are limited on our side for testing.

Thank you for the ping, I still have the device but it would need to be booted up.
I will ask a friend to boot and try, but could you provide here the ouput file for easier testing ?
Maybe one file per IP type ?

**Comment by j3n57h0m45** [comment]
`localhost:41412/security/blocklist?ipv4only`
```shell
:global CrowdSecBlockIP do={
  :local list "CrowdSec"
  :local address $1
  :local comment $2
  :local timeout $3
  onerror e in={ 
    /ip firewall address-list add list=$list address=$address comment=$comment timeout="$timeout"
  } do={
    /ip firewall address-list remove [ find list=$list address="$address" ]
    /ip firewall address-list add list=$list address=$address comment=$comment timeout="$timeout"
  }
}
$CrowdSecBlockIP x.x.x.x "crowdsecurity/http-open-proxy" 118h18m40s
$CrowdSecBlockIP y.y.y.y "crowdsecurity/ssh-bf" 78h2m22s
```
`localhost:41412/security/blocklist?ipv6only`
```shell
:global CrowdSecBlockIPv6 do={
  :local list "CrowdSec"
  :local address $1
  :local comment $2
  :local timeout $3
  onerror e in={ 
    /ipv6 firewall address-list add list=$list address=$address comment=$comment timeout="$timeout"
  } do={
    /ipv6 firewall address-list remove [ find list=$list address="$address" ]
    /ipv6 firewall address-list add list=$list address=$address comment=$comment timeout="$timeout"
  }
}
$CrowdSecBlockIPv6 xxxx:xxxx:xxxx:xxx:x:x:xxxx:xxxx "crowdsecurity/http-probing" 150h35m17s
$CrowdSecBlockIPv6 xxxx:xxxx:xxxx:xx::x "crowdsecurity/http-bad-user-agent" 166h59m59s
```

File: pkg/formatters/mikrotik/mikrotik.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	err = parsedTemplate.Execute(&script, data)
+	if err != nil {
+		http.Error(w, "Error executing template "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	// Remove empty lines and trim leading/trailing white spaces from the script
+	scriptString := strings.TrimSpace(strings.ReplaceAll(script.String(), "\n\n", "\n"))
+	script.Reset()
+	script.WriteString(scriptString)
```
Comment (LaurenceJJones) [inline pkg/formatters/mikrotik/mikrotik.go:86]: Do we need to write to scriptString if we already did `script.String()` two lines above?

File: pkg/formatters/mikrotik/mikrotik.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+	// Get the content length of the script
+	contentLength := len([]byte(script.String()))
```
Comment (LaurenceJJones) [inline pkg/formatters/mikrotik/mikrotik.go:89]: We can just `len(scriptString)` no?

**Comment by LaurenceJJones** [comment]
@j3n57h0m45 is it okay, if I do some general testing on perf and see if I can reduce some allocs? I can already push to the branch but want to make sure not stepping on your toes.

**Comment by LaurenceJJones** [comment]
@j3n57h0m45 Can you give it a whirl and let me know if I havent broken everything 😆

**Comment by LaurenceJJones** [comment]
Further testing has shown one issue which I may want to fix, since we write directly on the response writer if we encounter an issue with the template the buffer is not flush and the client gets a half written response with the error at the bottom of the current written buffer. Since the template now handles the newlines without needing a string we can allocate a buffer of bytes before writing it the client.

We may want to do this depending if microtik can handle a half written response with an http error code if not we should act on the above. Rely on your knowledge of this as I dont know how microtik handles error codes if it does?

**Comment by j3n57h0m45** [comment]
> Further testing has shown one issue which I may want to fix, since we write directly on the response writer if we encounter an issue with the template the buffer is not flush and the client gets a half written response with the error at the bottom of the current written buffer. Since the template now handles the newlines without needing a string we can allocate a buffer of bytes before writing it the client.
> 
> We may want to do this depending if microtik can handle a half written response with an http error code if not we should act on the above. Rely on your knowledge of this as I dont know how microtik handles error codes if it does?

So we would actually need to supply the script that fetches and executes the cs-blocklist-mirror code? Where would you want to provide that? Maybe in the Readme?

**Comment by LaurenceJJones** [comment]
> So we would actually need to supply the script that fetches and executes the cs-blocklist-mirror code? Where would you want to provide that? Maybe in the Readme?

If that what is happening I would suggest it, as I dont know how this all works 😕

**Comment by j3n57h0m45** [comment]
Maybe let's avoid this rabbit hole. Can we ensure that the script is fully generated prior to sending it to the client?

Otherwise we need some failed test cases to see how the router handles it.

**Comment by LaurenceJJones** [comment]
> Can we ensure that the script is fully generated prior to sending it to the client?

Yep, we can just write to a buffer and if err is not nil before then the response is the error response is sent as a single line.

**Comment by j3n57h0m45** [comment]
👍

Also any of the http error message will trip up the router. But then again this should not be in scope of this code but rather the script that fetches the list on the device.

**Comment by LaurenceJJones** [comment]
> 👍
> 
> Also any of the http error message will trip up the router. But then again this should not be in scope of this code but rather the script that fetches the list on the device.

Exactly, the script should check the response code and then bail if not 200

**Comment by j3n57h0m45** [comment]
So it might be feasible to provide a starter script to fetch the list for MikroTik devices in the Readme.

**Comment by j3n57h0m45** [comment]
Did we drop the ball? Is anything holding back this PR?

**Comment by LaurenceJJones** [comment]
> Did we drop the ball? Is anything holding back this PR?



> Did we drop the ball? Is anything holding back this PR?

I was waiting for confirmation that the changes I made work and there is some documentation for a script in the other PR.

**Comment by j3n57h0m45** [comment]
Just tested your changes without any problems. Also updated #99 with an import script.

**Review by LaurenceJJones — APPROVED** [review]

**Pull request merged**